segmentstore/server/src/main/java/io/pravega/segmentstore/server/containers/ContainerEventProcessorImpl.java

```diff
@@ -0,0 +1,457 @@
+         * processing iteration or there is no outstanding data to process.
+         */
+        private CompletableFuture<Void> getDelayedFutureIfNeeded() {
+            return failedIteration.get() || outstandingBytes.get() == 0 ? Futures.delayedFuture(iterationDelay, executor) :
+                    Futures.delayedFuture(Duration.ZERO, executor);
+        }
+
+        /**
+         * Returns a {@link CompletableFuture} that results from the execution of the following tasks for each EventProcessor:
+         * i) Read available data in the Segment, ii) deserialize at most {@link EventProcessorConfig#getMaxItemsAtOnce()},
+         * iii) execute the handler function for the {@link EventProcessor}, and iv) truncate the internal Segment.
+         */
+        private CompletableFuture<Void> processEvents() {
+            final Timer iterationTime = new Timer();
+            return readEvents()
+                    .thenComposeAsync(this::applyProcessorHandler, executor)
+                    .thenComposeAsync(readResult -> truncateInternalSegment(readResult, iterationTime), executor)
+                    .handleAsync((r, ex) -> {
+                        // If we got an exception different from NoDataAvailableException, report it as something is off.
+                        if (ex != null && !(Exceptions.unwrap(ex) instanceof NoDataAvailableException)) {
+                            log.warn("{}: Processing iteration failed, retrying.", this.traceObjectId, ex);
+                            failedIteration.set(true);
+                        } else {
+                            failedIteration.set(false);
+                        }
+                        return null;
+                    });
+        }
+
+        private CompletableFuture<EventsReadAndTruncationPoints> readEvents() {
+            AtomicLong startOffset = new AtomicLong(0L);
+            AtomicInteger readLength = new AtomicInteger(0);
+            return CompletableFuture.supplyAsync(() -> {
+                                        // Need to get the start and end offsets to properly create async read request.
+                                        startOffset.set(segment.getInfo().getStartOffset());
+                                        readLength.set((int) Math.min(segment.getInfo().getLength() - startOffset.get(), MAX_TOTAL_EVENT_SIZE));
+                                        return segment.read(startOffset.get(), readLength.get(), containerOperationTimeout);
+                                    }, executor)
+                                    .thenComposeAsync(rr -> AsyncReadResultProcessor.processAll(rr, executor, containerOperationTimeout), executor)
+                                    .thenApply(inputData -> deserializeEvents(inputData, startOffset.get(), readLength.get()));
+        }
+
+        private EventsReadAndTruncationPoints deserializeEvents(BufferView inputData, long startOffset, int maxLength) {
```

Comment:
I am not sure I am following how we deal with partial events.

Reply:
See reply above.